# Task 4.7 - Step 1: Exercise Follow Along

## Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## If Statements with User Defined Functions Small Practice

In [2]:
#Create a small test function

def add_num(x,y):
    sum = x + y
    return sum


In [3]:
#Run function

add_num(9,11)

20

## If Statements with Instacart

In [4]:
#Set path variable

path = r'C:\Users\HP\Documents\CareerFoundry\Data Immersion\Achievement 4\05-2023 Instacart Basket Analysis'

In [5]:
path

'C:\\Users\\HP\\Documents\\CareerFoundry\\Data Immersion\\Achievement 4\\05-2023 Instacart Basket Analysis'

In [6]:
#Import ords_prods_merged

df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [7]:
#Check if it imported correctly

df_ords_prods_merged

order_id user_id  order_number  orders_day_of_week  \
0         2539329       1             1                   2   
1         2398795       1             2                   3   
2          473747       1             3                   3   
3         2254736       1             4                   4   
4          431534       1             5                   4   
...           ...     ...           ...                 ...   
32404854  1320836  202557            17                   2   
32404855    31526  202557            18                   5   
32404856   758936  203436             1                   2   
32404857  2745165  203436             2                   3   
32404858  3093936  205420             1                   4   

          order_hour_of_day  days_since_last_order  product_id  \
0                         8                    NaN         196   
1                         7                   15.0         196   
2                        12                   21.0         196   
3                         7                   29.0         196   
4                        15                   28.0         196   
...                     ...                    ...         ...   
32404854                 15                    1.0       43553   
32404855                 11                    3.0       43553   
32404856                  7                    NaN       42338   
32404857                  5                   15.0       42338   
32404858                 14                    NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge  
0               77              7     9.0   both  
1               77              7     9.0   both  
2               77              7     9.0   both  
3               77              7     9.0   both  
4               77              7     9.0   both  
...            ...            ...     ...    ...  
32404854        64              7     3.7   both  
32404855        64              7     3.7   both  
32404856        50             19     6.9   both  
32404857        50             19     6.9   both  
32404858       130             14    10.3   both  

[32404859 rows x 14 columns]

In [8]:
df = df_ords_prods_merged[:1000000]

In [9]:
df

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0       2539329       1             1                   2                  8   
1       2398795       1             2                   3                  7   
2        473747       1             3                   3                 12   
3       2254736       1             4                   4                  7   
4        431534       1             5                   4                 15   
...         ...     ...           ...                 ...                ...   
999995  2836489  164627            14                   0                 15   
999996  1843600  164632             5                   1                 19   
999997   733106  164632             9                   2                 22   
999998  1650124  164632            17                   6                 17   
999999  3378258  164632            22                   6                 10   

        days_since_last_order  product_id  add_to_cart_order  reordered  \
0                         NaN         196                  1          0   
1                        15.0         196                  1          1   
2                        21.0         196                  1          1   
3                        29.0         196                  1          1   
4                        28.0         196                  1          1   
...                       ...         ...                ...        ...   
999995                    6.0       30489                  2          0   
999996                    9.0       30489                  2          0   
999997                   10.0       30489                  3          1   
999998                   13.0       30489                  1          1   
999999                   15.0       30489                  3          1   

           product_name  aisle_id  department_id  prices _merge  
0                  Soda        77              7     9.0   both  
1                  Soda        77              7     9.0   both  
2                  Soda        77              7     9.0   both  
3                  Soda        77              7     9.0   both  
4                  Soda        77              7     9.0   both  
...                 ...       ...            ...     ...    ...  
999995  Original Hummus        67             20     7.5   both  
999996  Original Hummus        67             20     7.5   both  
999997  Original Hummus        67             20     7.5   both  
999998  Original Hummus        67             20     7.5   both  
999999  Original Hummus        67             20     7.5   both  

[1000000 rows x 14 columns]

In [10]:
#Define the function

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [11]:
#Apply the functino

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\HP\AppData\Local\Temp\ipykernel_16164\4053350450.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [12]:
#Count the number of values

df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [13]:
#Find max value in the dataframe

df['prices'].max()

14.8

## If statements with the loc() Function

In [14]:
#First condition: High range if price over 15

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\HP\AppData\Local\Temp\ipykernel_16164\3310151462.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [15]:
#Second condition: Mid range if price is between 5 and 15

df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [16]:
#Third condition: Low range if price is under 5

df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
#Count values to see if it matches user-defined function without loc()

df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [18]:
#Use function on entire dataframe instead of subset

df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [19]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [20]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
#Count values

df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

## If Statements with For-Loops

In [22]:
#Make a for-loop for a simple function

for x in range(30,45):
    print('My age is %d'% (x))

My age is 30
My age is 31
My age is 32
My age is 33
My age is 34
My age is 35
My age is 36
My age is 37
My age is 38
My age is 39
My age is 40
My age is 41
My age is 42
My age is 43
My age is 44


### Make for loop to summarize how busy each day of the week is

In [23]:
# Count frequency of orders on each day of the week

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [24]:
#Function with the for-loop

result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [25]:
#Check the result

result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [26]:
#Set busiest_day column to the results of the for-loop

df_ords_prods_merged['busiest_day'] = result

In [27]:
#Count the busiest day values

df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

# Step 2: Create a new column that has the following criteria:
## "Busiest Days" matches the two busiest days of the week
## "Slowest Days" matches the two slowest days of the week
## "Regularly Busy" matches the others

In [28]:
#Check to see which days match the busiest days and slowest days

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [29]:
#Create new function with the for-loop

result_2 = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest Days")
  else:
    result_2.append("Regularly busy")

In [30]:
#Check the new result

result_2

['Regularly busy',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest Days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Slowest Days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest Days',
 'Regularly busy',
 'Slowest Days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest Days',
 'Slowest Days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Slowest Days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest 

In [31]:
#Set a new column for the results in the new for-loop

df_ords_prods_merged['busiest_days'] = result_2

# Step 3: Check the values of the new column for accuracy.

In [32]:
#Count the busiest_days values

df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest Days       7624336
Name: busiest_days, dtype: int64

### The total number of values is the same as the total number in the other "busiest_day" column (32404859).

# Step 4: Find the busiest hours of the day in groups with "Most orders", "Average orders" and "Fewest orders". Create a new column with these labels called "busiest_period_of_day".

In [33]:
#Look to see which column will be needed.

df_ords_prods_merged.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days  
0  Regularly busy  Regularly busy  
1  Regularly busy    Slowest Days  
2  Regularly busy    Slowest Days  
3      Least busy    Slowest Days  
4      Least busy    Slowest Days

In [34]:
# Use order_hour_of_day values. Check frequency of the order hours. 

df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

### The 24 listed hours can be split up into 3 groups, with the top 8 being "Most orders", middle 8 being "Average orders", and last 8 being "Fewest orders"

In [35]:
# New Function with the for-loop

result_3 = []

for value in df_ords_prods_merged["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [17, 8, 18, 19, 20, 7, 21, 22]:
    result_3.append("Average orders")
  else:
    result_3.append("Fewest orders")

In [36]:
# Check the new result

result_3

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

In [37]:
#Set a new column for the results in the new for-loop

df_ords_prods_merged['busiest_period_of_day'] = result_3

In [38]:
#Check to see if column was inserted properly

df_ords_prods_merged.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest Days        Average orders  
2  Regularly busy    Slowest Days           Most orders  
3      Least busy    Slowest Days        Average orders  
4      Least busy    Slowest Days           Most orders

# Step 5: Print the frequency for this new column.

In [39]:
#Check frequency

df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

### Once again, the total number of orders matches the total amount of rows (32404859).

# Step 7: Export your dataframe as a pickle file

In [40]:
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated.pkl'))